nixos/modules/services/desktop-managers/cosmic.nix

```
@@ -0,0 +1,173 @@
+# SPDX-License-Identifier: MIT
+# SPDX-FileCopyrightText: Lily Foster <lily@lily.flowers>
+# Portions of this code are adapted from nixos-cosmic
+# https://github.com/lilyinstarlight/nixos-cosmic
+
+{
+  config,
+  lib,
+  pkgs,
+  utils,
+  ...
+}:
+
+let
+  cfg = config.services.desktopManager.cosmic;
+in
+{
+  meta.maintainers = with lib.maintainers; [
+    thefossguy
+    HeitorAugustoLN
+  ];
+
+  options = {
+    services.desktopManager.cosmic = {
+      enable = lib.mkEnableOption "Enable the COSMIC desktop environment";
+
+      xwayland.enable = lib.mkEnableOption "Xwayland support for `cosmic-comp`" // {
+        default = true;
+      };
+    };
+
+    environment.cosmic.excludePackages = lib.mkOption {
+      description = "List of COSMIC packages to exclude from the default environment";
+      type = lib.types.listOf lib.types.package;
+      default = [ ];
+      example = lib.literalExpression "[ pkgs.cosmic-player ]";
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    # Environment packages
+    environment.pathsToLink = [
+      "/share/backgrounds"
+      "/share/cosmic"
+    ];
+    environment.systemPackages = utils.removePackagesByName (
+      with pkgs;
+      [
+        adwaita-icon-theme
+        alsa-utils
+        cosmic-applets
+        cosmic-applibrary
+        cosmic-bg
+        (cosmic-comp.override { useXWayland = cfg.xwayland.enable; })
```

This override should be set to `false`, as the module already includes `xwayland` in the system packages. There's no need to add it again for `cosmic-comp` specifically.

```suggestion
        (cosmic-comp.override { useXWayland = false; })
```

---

No, it doesn't disable XWayland support. It's already handled below by adding it to the system packages when `xwayland.enable` is set to `true`.

---

Yeah, but we're removing xwayland from `cosmic-comp`'s `$PATH`. I have no idea how that affects xwayland in the cosmic DE. Let's get this through and we can test it out?

---

Setting `useXWayland = false` in the cosmic-comp override doesn't disable XWayland support for the COSMIC desktop environment, but rather, it prevents PATH pollution by ensuring XWayland is provided through a single, explicit source in the system packages instead of being duplicated within the compositor package itself. When a user enables XWayland with `services.desktopManager.cosmic.xwayland.enable = true`, the module correctly adds the standalone `xwayland` package to the system packages list via `lib.optionals cfg.xwayland.enable [ xwayland ]`, providing XWayland through the system-wide PATH rather than redundantly including it in cosmic-comp's wrapped PATH environment.

---

Oh okay, so we should modify the `cosmic-comp` package to always disable xwayland support because we will toggle xwayland support using the cosmic module.

---

> Oh okay, so we should modify the `cosmic-comp` package to always disable xwayland support because we will toggle xwayland support using the cosmic module.

No, this is the correct approach, the package should have reasonable defaults for standalone use, while the module makes the conscious decision to override it.